cub/device/dispatch/dispatch_histogram.cuh

```diff
@@ -237,79 +241,56 @@ struct DispatchHistogram
     // Scales samples to evenly-spaced bins
     struct ScaleTransform
     {
+        using CommonT = typename cuda::std::common_type<LevelT, SampleT>::type;
```

Comment:
I think this is a breaking change. Requirements on types used to be: 
```cpp
template<typename L>
concept Layout = requires(L&& l, L &&r)
{
    requires std::is_trivially_copyable_v<L>;
    { l >= r } -> std::convertible_to<bool>;
    { l < r } -> std::convertible_to<bool>;
    { (l - r) / r } -> std::convertible_to<int>;
};

template<typename S, typename L>
concept Sample = Layout<L> && requires
{
    requires std::is_convertible_v<S, L>;
    requires std::is_copy_constructible_v<S>;
};

template <Layout L, Sample<L> ST>
void histogram(int num_levels, 
               L min,
               L max,
               ST* samples);
```

Now they are something along the following lines:
```cpp
template<typename L, typename R>
concept Common = requires(
    std::common_type_t<L, R> l, 
    std::common_type_t<L, R> r)
{
    requires std::is_convertible_v<int, std::common_type_t<L, R>>;
    typename std::common_type<L, R>::type;
    { l >= r } -> std::convertible_to<bool>;
    { l < r } -> std::convertible_to<bool>;
    { (l - r) / r } -> std::convertible_to<int>;
};

template<typename L, typename CommonT>
concept Layout = requires
{
    requires std::is_trivially_copyable_v<L>;
    requires std::is_convertible_v<L, CommonT>;
};

template<typename S, typename CommonT>
concept Sample = requires
{
    requires std::is_copy_constructible_v<S>;
    requires std::is_convertible_v<S, CommonT>;
};
```

The new constrain `std::is_convertible_v<int, std::common_type_t<L, R>>` is caused by `static_cast<CommonT>(num_output_levels - 1)`. Besides that, a different issue might be found in the following use case:

```cpp
class custom_layout_t
{
  int val{};
public:
  explicit custom_layout_t(int v) : val(v) {}

  operator int() {
    return val;
  }

  // Methods ignored
  custom_layout_t operator-(const custom_layout_t &rhs) { return custom_layout_t{val - rhs.val}; }
  bool operator>=(const custom_layout_t &rhs) { return val >= rhs.val; }
  bool operator<(const custom_layout_t &rhs) { return val < rhs.val; }
  int operator/(const custom_layout_t &rhs) { return val / rhs.val; }
};

int main() {
  custom_layout_t min {0};
  custom_layout_t max {2};

  int samples[2] { 0, 1 };

  int num_levels = 2;

  histogram(num_levels, min, max, samples);
}
```

Before this changes, we'd use `custom_layout_t` methods, whereas now, we'd use integer arithmetic (since the common type is `int`). I think there might be other subtle differences in the requirements and/or behavior. 

My suggestion would be to start with the documentation and provide some constrains on the types. Do we expect the algorithm to work with user types at all?

Reply:
Integer arithmetics is not a good idea at all, see https://github.com/NVIDIA/cub/issues/489

Reply:
Added compile time checks for the `common_type` and `is_convertible` requirements and updated documentation to explicitly specify these.

Also updated the logic to fix the round issue reported in #489.